README.md
# Introduction

[![CircleCI](https://circleci.com/gh/AumitLeon/archMLP.svg?style=svg)](https://circleci.com/gh/AumitLeon/archMLP) [![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release) [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)
[![codecov](https://codecov.io/gh/AumitLeon/archMLP/branch/master/graph/badge.svg)](https://codecov.io/gh/AumitLeon/archMLP)


ArchMLP is a Machine Learning pipeline that creates an analytical setting and operational setting. The analytical setting focuses on model development and enhancement, while the operational setting is focused on model deployment. ArchMLP is designed as a microservices architecture via a Docker containerization schema.

ArchMLP \(Architect Machine Learning Platform\) is a scalable machine learning platform designed with a microservices architecture. The platform was originally theorized in this [work](https://aumitleon.com/assets/projects/archMLP_report.pdf) during a senior seminar course at Middlebury College. 

This platform combines the _analytical_ and the _operational_ settings involved in developing modern machine learning models. The _analytical_ setting is concerned with developing models and attaining a high test/validation accuracy by reducing overfitting. The _operational_ setting seeks to optimize performance by making the model queryable and highly available for users.





CHANGELOG.md
## [1.1.1](https://github.com/AumitLeon/archMLP/compare/v1.1.0...v1.1.1) (2019-12-25)


### Bug Fixes

* **semantic-release:** update semantic release plugins ([40e2190](https://github.com/AumitLeon/archMLP/commit/40e2190de88168ad503f451bc7b1efc1c2451b8d))


server/server.test.js
const { app } = require('./server');
const path = require('path');
const request = require('supertest');
const rimraf = require('rimraf');

test('Server "smoke" test', () => {
  expect(app).toBeDefined();
});

describe('Upload dataset endpoint ', () => {
  const filePath = path.resolve(__dirname, 'testFiles/test.csv');
  const errorFilePath = path.resolve(__dirname, 'testFiles/test.txt');

  afterAll(() => {
    // Delete uploaded test files
    // rimraf requires a callback, here we provide an empty
    rimraf(path.resolve(__dirname, 'uploads/'), () => {});
  });

  test('POST /api/v1/uploadData - upload a new CSV file successfully', () => {
    return request(app)
      .post('/api/v1/uploadData')
      .attach('file', filePath)
      .then(response => {
        expect(response.status).toBe(200);
        expect(response.body.error).toBeFalsy();
        expect(response.body.success).toBe('File successfully uploaded!');
      });
  });

  test('POST /api/v1/uploadData - Upload fails', () => {
    return request(app)
      .post('/api/v1/uploadData')
      .then(response => {
        expect(response.status).toBe(400);
        expect(response.body.success).toBeFalsy();
        expect(response.body.error).toBe('File required!');
      });
  });

  test('POST /api/v1/uploadData - Upload fails with wrong type of file', () => {
    return request(app)
      .post('/api/v1/uploadData')
      .attach('file', errorFilePath)
      .then(response => {
        expect(response.status).toBe(400);
        expect(response.body.success).toBeFalsy();
      });
  });
});

describe('Upload dataset name endpoint', () => {
  const datasetName = { name: 'test' };
  test('POST /api/v1/setDataName - recieve a new dataset name', () => {
    return request(app)
      .post('/api/v1/setDataName')
      .send(datasetName)
      .then(response => {
        expect(response.status).toBe(200);
        expect(response.body.error).toBeFalsy();
        expect(response.body.success).toBe('Dataset name has been received!');
      });
  });

  test('POST /api/v1/setDataName - 400 response with empty body', () => {
    return request(app)
      .post('/api/v1/setDataName')
      .then(response => {
        expect(response.status).toBe(400);
        expect(response.body.success).toBeFalsy();
        expect(response.body.error).toBe('Failed to receive dataset name.');
      });
  });

  test('POST /api/v1/setDataName - 400 response with name containing non-alphanumerics', () => {
    return request(app)
      .post('/api/v1/setDataName')
      .send({ name: 'abcd$' })
      .then(response => {
        expect(response.status).toBe(400);
        expect(response.body.success).toBeFalsy();
        expect(response.body.error).toBe(
          'Dataset name can only contain alphanumeric characters.'
        );
      });
  });

  test('POST /api/v1/setDataName - 400 response with name of length > 31 chars', () => {
    return request(app)
      .post('/api/v1/setDataName')
      .send({ name: 'abcdefghijklmnopqrstuvwxyz123456' })
      .then(response => {
        expect(response.status).toBe(400);
        expect(response.body.success).toBeFalsy();
        expect(response.body.error).toBe(
          'Dataset name can contain at-most 31 characters.'
        );
      });
  });
});


client/src/apiCalls.test.js
import { upload, setName } from './apiCalls';

describe('Test client side API calls', () => {
  beforeEach(() => {
    fetch.resetMocks();
  });

  test('Calls /api/v1/uploadData and returns a successful response', () => {
    fetch.mockResponseOnce(JSON.stringify({ data: '12345' }));

    //assert on the response
    upload({}).then(res => {
      expect(res.body.success).toEqual('12345');
    });
  });

  test('Calls /api/v1/setDataName and returns a successful response', () => {
    fetch.mockResponseOnce(JSON.stringify({ name: '12345' }));
    //assert on the response
    setName({}).then(res => {
      expect(res.body.success).toEqual('12345');
    });
  });
});


client/src/components/FileUpload.test.js
import React from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import FileUpload from './FileUpload';
import { createShallow, createMount } from '@material-ui/core/test-utils';

const props = {
  dataName: '',
  setDataName: jest.fn(),
  file: {},
  setFile: jest.fn()
};

describe('FileUpload tests', () => {
  let wrapper;
  let shallow;

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    wrapper = shallow(<FileUpload {...props} />);
  });

  test('Renders form', () => {
    expect(wrapper.find('form')).toHaveLength(1);
  });

  test('Renders dataset name input', () => {
    expect(wrapper.find(TextField)).toHaveLength(1);
  });

  test('Renders upload button', () => {
    expect(wrapper.find(Button)).toHaveLength(1);
  });

  test('Render help text for dataset input field', () => {
    const mount = createMount();
    const helper = mount(<FileUpload {...props} />);

    expect(helper.find(TextField).text()).toEqual('Dataset Name *​');
  });

  test('Updating dataset name calls setDataName callback and sets empty helper text', () => {
    wrapper
      .find(TextField)
      .at(0)
      .simulate('change', { target: { value: 'newName' } });
    expect(props.setDataName).toHaveBeenCalled();
    expect(wrapper.find(TextField).props().error).toEqual(false);
    expect(wrapper.find(TextField).props().helperText).toEqual('');
  });

  test('Updating dataset name calls setFile callback', () => {
    wrapper
      .find('input')
      .simulate('change', { target: { files: ['dummy.value'] } });
    expect(props.setFile).toHaveBeenCalled();
  });

  test('Proper helper text on name with length > 31', () => {
    wrapper
      .find(TextField)
      .at(0)
      .simulate('change', {
        target: { value: 'abcdefghijklmnopqrstuvwxyz123456' }
      });
    expect(wrapper.find(TextField).props().helperText).toEqual(
      'Dataset name can only contain 31 characters'
    );
    // Disable button on invalid name input
    expect(wrapper.find(TextField).props().error).toEqual(true);
  });

  test('Proper helper text on name with non-alphanumerics', () => {
    wrapper
      .find(TextField)
      .at(0)
      .simulate('change', { target: { value: 'abcd$' } });
    expect(wrapper.find(TextField).props().helperText).toEqual(
      'Dataset name can only contain alphanumerics'
    );
    // Disable button on invalid name input
    expect(wrapper.find(TextField).props().error).toEqual(true);
  });

  test('Proper helper text on empty input', () => {
    wrapper
      .find(TextField)
      .at(0)
      .simulate('change', { target: { value: '' } });
    expect(wrapper.find(TextField).props().helperText).toEqual('');
    expect(wrapper.find(TextField).props().error).toEqual(false);
  });
});


client/src/App.js
import React, { useState } from 'react';
import './App.css';
import FileUpload from './components/FileUpload';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import { ThemeProvider, createMuiTheme } from '@material-ui/core/styles';
import { Helmet, HelmetProvider } from 'react-helmet-async';
import { blue } from '@material-ui/core/colors';

function App() {
  const [dataName, setDataName] = useState('');
  const [file, setFile] = useState({});

  // Set theme based on user preferences (from OS)
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)');
  const theme = React.useMemo(
    () =>
      createMuiTheme({
        palette: {
          primary: blue,
          accent: blue[200],
          type: prefersDarkMode ? 'dark' : 'light'
        }
      }),
    [prefersDarkMode]
  );

  return (
    <div className="App">
      <HelmetProvider>
        <Helmet title="ArchMLP" />
      </HelmetProvider>
      <ThemeProvider theme={theme}>
        <FileUpload
          dataName={dataName}
          setDataName={setDataName}
          file={file}
          setFile={setFile}
        />
      </ThemeProvider>
    </div>
  );
}

export default App;


client/src/apiCalls.js
// Upload the dataset
export const upload = file => {
  const formData = new FormData();
  formData.append('file', file);
  return fetch('/api/v1/uploadData', {
    // Your POST endpoint
    method: 'POST',
    body: formData // This is your file object
  })
    .then(
      response => response.json() // if the response is a JSON object
    )
    .then(
      success => console.log(success) // Handle the success response object
    )
    .catch(
      error => console.log(error) // Handle the error response object
    );
};

// Set dataset name
export const setName = dataName => {
  return fetch('/api/v1/setDataName', {
    // Your POST endpoint
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name: dataName })
  })
    .then(
      response => response.json() // if the response is a JSON object
    )
    .then(
      success => console.log(success) // Handle the success response object
    )
    .catch(
      error => console.log(error) // Handle the error response object
    );
};


client/src/components/FileUpload.js
import React, { useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import CloudUploadIcon from '@material-ui/icons/CloudUpload';
import Avatar from '@material-ui/core/Avatar';
import CssBaseline from '@material-ui/core/CssBaseline';
import TextField from '@material-ui/core/TextField';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import Box from '@material-ui/core/Box';
import Container from '@material-ui/core/Container';
import UploadDiaglog from './UploadDiaglog';
import PropTypes from 'prop-types';
import Tooltip from '@material-ui/core/Tooltip';

const useStyles = makeStyles(theme => ({
  input: {
    display: 'none'
  },
  paper: {
    marginTop: theme.spacing(8),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center'
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(1)
  },
  submit: {
    margin: theme.spacing(3, 0, 2)
  }
}));

function Copyright() {
  return (
    <Typography variant="body2" color="textSecondary" align="center">
      {'Copyright © '}
      <Link color="inherit" href="https://aumitleon.com/">
        Aumit Leon
      </Link>{' '}
      {new Date().getFullYear()}
      {'.'}
    </Typography>
  );
}

function FileUpload({ dataName, setDataName, file, setFile }) {
  const classes = useStyles();

  // State variables for UploadDialog
  const [open, setOpen] = useState(false);
  const [nameError, setNameError] = useState('');

  const handleClickOpen = uploadedFile => {
    setFile(uploadedFile);
    setOpen(true);
  };

  // Only accept alphanermic strings with length <= 31
  // Set text error accordingly
  const validateInput = name => {
    if (name.match(/^[a-z0-9]+$/i) && name.length <= 31) {
      setDataName(name);
      setNameError('');
    } else if (name === '') {
      setNameError('');
    } else if (name.match(/^[a-z0-9]+$/i) === null) {
      setNameError('Dataset name can only contain alphanumerics');
    } else if (name.length > 31) {
      setNameError('Dataset name can only contain 31 characters');
    }
  };

  return (
    <div>
      <AppBar position="static" color="primary">
        <Toolbar>
          <Typography variant="h5" className={classes.title}>
            ArchMLP
          </Typography>
        </Toolbar>
      </AppBar>
      <Container component="main" maxWidth="xs">
        <CssBaseline />
        <div className={classes.paper}>
          <Avatar className={classes.avatar}>
            <CloudUploadIcon />
          </Avatar>
          <Typography component="h1" variant="h5">
            Upload Dataset
          </Typography>
          <form className={classes.form} noValidate>
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              id="dataset"
              label="Dataset Name"
              name="dataset"
              autoComplete="dataset"
              helperText={nameError}
              autoFocus
              error={nameError === '' ? false : true}
              onChange={event => validateInput(event.target.value)}
            />
            <input
              accept=".csv"
              className={classes.input}
              id="contained-button-file"
              type="file"
              onChange={event => handleClickOpen(event.target.files[0])}
            />
            <label htmlFor="contained-button-file">
              <Tooltip title="File must be CSV format" placement="right">
                <Button
                  fullWidth
                  variant="contained"
                  color="primary"
                  component="span"
                  className={classes.submit}
                  disabled={nameError === '' ? false : true}
                >
                  Upload
                </Button>
              </Tooltip>
            </label>
            {file && (
              <UploadDiaglog
                open={open}
                setOpen={setOpen}
                dataName={dataName}
                file={file}
              />
            )}
            <Grid container>
              <Grid item xs>
                <Link
                  href="https://aumitleon1.gitbook.io/archmlp/"
                  variant="body2"
                >
                  Documentation
                </Link>
              </Grid>
            </Grid>
          </form>
        </div>
        <Box mt={8}>
          <Copyright />
        </Box>
      </Container>
    </div>
  );
}

FileUpload.propTypes = {
  dataName: PropTypes.string.isRequired,
  setDataName: PropTypes.func.isRequired,
  file: PropTypes.object.isRequired,
  setFile: PropTypes.func.isRequired
};

export default FileUpload;


client/src/components/UploadDiaglog.js
import React from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import PropTypes from 'prop-types';
import { upload, setName } from '../apiCalls';

function UploadDialog({ open, setOpen, dataName, file }) {
  //console.log(file.name)
  const handleClose = () => {
    setOpen(false);
  };
  const handleUpload = () => {
    upload(file); // Upload the dataset
    setName(dataName); // Set the dataset name
    setOpen(false); // Close the dialog
  };
  return (
    <Dialog
      open={open}
      onClose={handleClose}
      aria-labelledby="alert-dialog-title"
      aria-describedby="alert-dialog-description"
    >
      <DialogTitle id="alert-dialog-title">{'Upload this file?'}</DialogTitle>
      <DialogContent>
        <DialogContentText id="alert-dialog-description">
          {`The dataset you upload will be used to preprocess, train, and deploy your machine learning model. 
          Are you sure you want to use the file ${file.name} with the name ${dataName}?`}
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={handleClose} color="primary">
          Cancel
        </Button>
        {/* Pass a handler to make POST request to API */}
        <Button onClick={handleUpload} color="primary" autoFocus>
          Upload
        </Button>
      </DialogActions>
    </Dialog>
  );
}

UploadDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  setOpen: PropTypes.func.isRequired,
  dataName: PropTypes.string.isRequired,
  file: PropTypes.object.isRequired
};

export default UploadDialog;


server/server.js
const express = require('express');
const path = require('path'); // eslint-disable-line global-require
const multer = require('multer');
const mkdirp = require('mkdirp');
const rateLimit = require('express-rate-limit');
const morgan = require('morgan');
const rfs = require('rotating-file-stream');
const log4js = require('log4js');
const bodyParser = require('body-parser');

// Node logger config
log4js.configure({
  appenders: { console: { type: 'file', filename: 'logs/console.log' } },
  categories: { default: { appenders: ['console'], level: 'debug' } }
});
const logger = log4js.getLogger('console');

// multer config
const UPLOAD_PATH = 'uploads';
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    const dir = `${UPLOAD_PATH}/`;
    mkdirp(dir, err => cb(err, dir));
  },
  filename: (req, file, cb) => {
    // Remove extension, add time stamp, readd extension
    cb(null, `${file.originalname.slice(0, -4)}-${Date.now()}.csv`);
  }
});

const upload = multer({
  storage,
  fileFilter: (req, file, cb) => {
    const filetypes = /csv/;
    const mimetype = filetypes.test(file.mimetype);
    const extname = filetypes.test(
      path.extname(file.originalname).toLowerCase()
    );
    logger.info('Attempting to validate file');
    // Successful validation
    if (mimetype && extname) {
      logger.info('File validation successful.');
      return cb(null, true);
    }
    // File validation error
    logger.error('File validation failed.');
    req.fileValidationError = `File upload only supports the following filetypes - ${filetypes}`;
    return cb(null, false, req.fileValidationError);
  }
}).single('file');

// Rate limiting config
// Enable if you're behind a reverse proxy (Heroku, Bluemix, AWS ELB, Nginx, etc)
// see https://expressjs.com/en/guide/behind-proxies.html
// app.set('trust proxy', 1);
const apiLimiter = rateLimit({
  windowMs: 60 * 60 * 1000, // 15 minutes
  max: 50 // 50 requests per IP within 15 minute window
});

// logging config
// create a rotating write stream
const accessLogStream = rfs.createStream('access.log', {
  interval: '1d', // rotate daily
  path: path.join(__dirname, 'logs')
});

// Resolve client build directory as absolute path to avoid errors in express
const buildPath = path.resolve(__dirname, '../client/build');

const app = express();

// Express only serves static assets in production
if (process.env.NODE_ENV === 'production') {
  // Serve any static files as first priority
  app.use(express.static(buildPath));
}

// Middleware
// rate limiting middleware
// only apply to requests that begin with /api/
app.use('/api/v1/uploadData', apiLimiter);

// setup the logger
app.use(morgan('combined', { stream: accessLogStream }));

// body parser middleware
app.use(bodyParser.json());

// State variables
let dataName; // dataset name
let file; // uploaded dataset

// POST route for uploading new file using upload (multer middleware)
app.post('/api/v1/uploadData', upload, (req, res, next) => {
  logger.info('User requested /api/v1/uploadData');
  try {
    if (req.fileValidationError) {
      logger.error(req.fileValidationError);
      res.status(400).send({ error: req.fileValidationError });
      next(req.fileValidationError);
    } else {
      logger.info(`File ${req.file.originalname} uploaded successfuly.`);
      file = req.file;
      logger.info(`File object: ${file}`);
      res.status(200).send({ success: 'File successfully uploaded!' });
    }
  } catch (err) {
    logger.error('File missing from request.');
    res.status(400).send({ error: 'File required!' });
  }
});

// POST route for setting dataset name
app.post('/api/v1/setDataName', (req, res) => {
  logger.info('User requested /api/v1/setDataName');
  if (req.body.name) {
    logger.info(`Dataset name ${req.body.name} received.`);
    if (req.body.name.match(/^[a-z0-9]+$/i) && req.body.name.length <= 31) {
      dataName = req.body.name;
      logger.info(`Dataset name ${dataName} validation successful.`);
      res.status(200).send({ success: 'Dataset name has been received!' });
    } else if (req.body.name.match(/^[a-z0-9]+$/i) === null) {
      logger.error(`Dataset name ${dataName} has non-alphanumeric characters.`);
      res.status(400).send({
        error: 'Dataset name can only contain alphanumeric characters.'
      });
    } else if (req.body.name.length > 31) {
      logger.error(`Dataset name ${dataName} has length greater than 31.`);
      res
        .status(400)
        .send({ error: 'Dataset name can contain at-most 31 characters.' });
    }
  } else {
    logger.error('Dataset name not received');
    res.status(400).send({ error: 'Failed to receive dataset name.' });
  }
});

// Express only serves static assets in production
if (process.env.NODE_ENV === 'production') {
  // All remaining requests return the React app, so it can handle routing.
  app.get('*', (request, response) => {
    response.sendFile(path.join(buildPath, 'index.html'));
  });
}

module.exports = {
  app
};
